pkg/controller/secretsyncer/secretsyncer.go

```diff
@@ -0,0 +1,205 @@
+	}
+	log.WithField("request", sourceSecretName.String()).Trace("Reconciling")
+
+	source := &corev1.Secret{}
+	if err := r.referenceClusterClient.Get(r.ctx, sourceSecretName, source); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get secret %s: %w", sourceSecretName.String(), err)
+	}
+	if source.DeletionTimestamp != nil {
+		log.Debug("not doing work for secret because it is being deleted")
+		return nil
+	}
+
+	if len(source.Data) == 0 {
+		log.Trace("not updating target secret as source has no data")
+		return nil
+	}
+
+	var mirrorErrors []error
+	for _, mirrorConfig := range r.config().Secrets {
+		if mirrorConfig.From.Namespace != sourceSecretName.Namespace || mirrorConfig.From.Name != sourceSecretName.Name {
+			continue
+		}
+		// TODO: Ideally we would would create one reconcile.Request per target. This here means
+		// that if a single target is broken, it will send all targets into exponential backoff.
+		mirrorErrors = append(mirrorErrors, r.mirrorSecret(log, client, source, mirrorConfig.To))
+	}
+
+	return utilerrors.NewAggregate(mirrorErrors)
+}
+
+func (r *reconciler) mirrorSecret(log *logrus.Entry, c ctrlruntimeclient.Client, source *corev1.Secret, to config.SecretLocation) error {
+	*log = *log.WithFields(logrus.Fields{"target-namespace": to.Namespace, "target-secret": to.Name})
+
+	secret, mutateFn := secret(
+		types.NamespacedName{Namespace: to.Namespace, Name: to.Name},
+		source.Data,
+		source.Type,
+	)
+	result, err := crcontrollerutil.CreateOrUpdate(r.ctx, c, secret, mutateFn)
+	if err != nil {
+		// Happens if someone changes the type, just re-create it
+		if strings.Contains(err.Error(), "field is immutable") {
+			if err := recreateSecret(r.ctx, c, secret); err != nil {
+				return fmt.Errorf("failed to recreate secret %s/%s: %w", secret.Namespace, secret.Name, err)
+			}
+		}
+		return fmt.Errorf("failed to upsert secret: %w", err)
+	}
+	if result != crcontrollerutil.OperationResultNone {
+		log.WithField("operation", result).Info("Upsert succeeded")
+	}
+	return nil
+}
+
+func recreateSecret(ctx context.Context, c ctrlruntimeclient.Client, s *corev1.Secret) error {
+	if err := c.Delete(ctx, s.DeepCopy()); err != nil {
+		return fmt.Errorf("failed to delete secret: %w", err)
+	}
+	if err := c.Create(ctx, s); err != nil {
+		return fmt.Errorf("failed to create secret: %w", err)
+	}
+	return nil
+}
+
+func secret(nn types.NamespacedName, data map[string][]byte, tp corev1.SecretType) (*corev1.Secret, crcontrollerutil.MutateFn) {
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: nn.Namespace,
+			Name:      nn.Name,
+		},
+	}
+	return s, func() error {
+		s.Data = data
+		s.Type = tp
+		return nil
+	}
+}
+
+func requestForCluster(clusterName, namespace, name string) reconcile.Request {
+	return reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: clusterName + "_" + namespace,
```

Comment:
I guess this is "The reconcile.Requests must include the cluster name":

Probably the framework did not consider `ClusteredNamespacedName`?
Or any place in the request to save additional info, like clusterName in our case?

Reply:
Yeah, the library wasn't made for multicluster controllers, you need to build certain things like this yourself.